File: tools/create-institute-config.pl
#!/usr/bin/perl
# Copyright 2018 Peter Varkoly <peter@varkoly.de>  Germany, Nuremberg

use strict;
use Data::Dumper;
use JSON::XS;
use Net::Netmask;
use Encode qw(encode decode);
binmode STDIN,  ":encoding(UTF-8)";
binmode STDOUT, ":encoding(UTF-8)";
binmode STDERR, ":encoding(UTF-8)";
use utf8;

my $Defaults = "/usr/share/cephalix/templates/Defaults.ini";

my @TO_CLEAN   = qw(
        uuid
        name
        type
        domain
        adminPW
);

my $READONLY   = {
        "CEPHALIX" => 1,
        "CEPHALIX_PATH"    => 1,
        "CEPHALIX_DOMAIN"  => 1,
	"CCODE" => 1,
	"LANGUAGE" => 1,
	"NTP" => 1,
        "ZADMIN"   => 1
};

my @SSL = qw(
        REPLACE-SSHKEY
        REPLACE-CA-CERT
        REPLACE-VPN-CERT
        REPLACE-VPN-KEY
        REPLACE-SERVER-KEY
        REPLACE-SERVER-CERT
        REPLACE-ADMIN-CERT
        REPLACE-ADMIN-KEY
        REPLACE-SCHOOL-KEY
        REPLACE-SCHOOL-CERT
);

my %SSLVARS = ();

my $hash = "";

#############################################
# Some helper functions

sub hash_to_json($) {
    my $hash = shift;
    my $json = '{';
    foreach my $key ( keys %{$hash} ) {
        my $value = $hash->{$key};
        $json .= '"'.$key.'":';
        if( $value eq 'true' ) {
       $json .= 'true,';
        } elsif ( $value eq 'false' ) {
       $json .= 'false,';
        } elsif ( $value =~ /^\d+$/ ) {
       $json .= $value.',';
        } else {
                $value =~ s/"/\\"/g;
                $json .= '"'.$value.'",';
        }
    }
    $json =~ s/,$//;
    $json .= '}';
}

sub convert_json_to_hash($)
{
   my $res = shift;
   $res =~ s/:false/:0/g;
   $res =~ s/:true/:1/g;
   $res =~ s/:null/:""/g;
   $res =~ s/":/"=>/g;
   return eval($res);
}


sub contains {
    my $a = shift;
    my $b = shift;
    foreach(@{$b}){
        if($a eq $_) {
                return 1;
        }
    }
    return 0;
}

sub write_file($$) {
  my $file = shift;
  my $out  = shift;
  local *F;
  open F, ">$file" || die "Couldn't open file '$file' for writing: $!; aborting";
  binmode F, ':encoding(utf8)';
  local $/ unless wantarray;
  print F $out;
  close F;
}

sub get_file($) {
    my $file     = shift;
    return undef if( ! -e $file );
    my $content  = '';
    local *F;
    open F, $file || return undef;
    binmode F, ':encoding(utf8)';
    while( <F> )
    {
       $content .= $_;
    }
    return $content;
}


#
#############################################

while(<STDIN>) {
   $hash .= $_;
}
my $TASK      =`uuidgen -t`; chomp $TASK;
system("mkdir -p /var/adm/oss/opentasks/");
write_file("/var/adm/oss/opentasks/$TASK",$hash);

my $reply = convert_json_to_hash($hash);

my $default  = decode_json(`cat $Defaults`);
my $CEPHALIX_PATH   = $default->{'CEPHALIX_PATH'};
my $CEPHALIX_DOMAIN = $default->{'CEPHALIX_DOMAIN'};
my $SAVE_NEXT = ( $default->{'SAVE_NEXT'} eq "yes" ) ? 1 : 0;
my $path = $CEPHALIX_PATH.'/configs/'.$reply->{"uuid"}.".ini";
my $AY_TEMPLATE = $reply->{"ayTemplate"} || "default";
my $XMLFile  = "/usr/share/cephalix/templates/autoyast-".$AY_TEMPLATE.".xml";

system("mkdir -p $CEPHALIX_PATH/configs/");

#Add the read only variables to the reply 
foreach my $k ( keys(%$READONLY) )
{
       $reply->{$k} = $default->{$k};
}

#Handle VPN
if( $reply->{'ipAdmin'} ne $reply->{'ipVPN'} ) {
	my $vpnblock = new Net::Netmask($reply->{'ipVPN'}."/255.255.255.252");
	my $VPN      = "ifconfig-push ".$vpnblock->nth(2).' '.$vpnblock->nth(1)."\n";
	$reply->{'ipVPN'} = $vpnblock->nth(2);
	if( $reply->{'fullrouting'} )
	{
		my $block = new Net::Netmask($reply->{"network"});
		$VPN .= 'iroute '.$block->base().' '.$block->mask()."\n";
	}
	write_file('/etc/openvpn/ccd/'.$reply->{'uuid'},$VPN);

	my $next     = $vpnblock->next();
	$default->{'ipVPN'} = $next;
}
if( $SAVE_NEXT )
{
        my $block = new Net::Netmask($reply->{"network"});
        my $next  = $block->next();
        $default->{'network'} = $next;
        my ( $a,$b,$c,$d ) = split /\./, $next;
        $default->{'ipAdmin'}	= "$a.$b.$c.".($d+2);
        $default->{'ipMail'}	= "$a.$b.$c.".($d+3);
        $default->{'ipPrint'}	= "$a.$b.$c.".($d+4);
        $default->{'ipProxy'}	= "$a.$b.$c.".($d+5);
        $default->{'ipBackup'}	= "$a.$b.$c.".($d+6);
        $default->{'anonDhcp'}	= "$a.$b.".($c+1).".0 $a.$b.".($c+1).".31";
        $default->{'firstRoom'}	= "$a.$b.".($c+2).".0";
}

foreach my $v ( @TO_CLEAN )
{
        $default->{$v} = '';
}
if( ! -e "$CEPHALIX_PATH/CA_MGM/certs/admin.".$reply->{"domain"}.".key.pem" ) {
        my $command = "$CEPHALIX_PATH/create_server_certificates.sh -P $CEPHALIX_PATH -O '".$reply->{'name'};
	if( defined $reply->{'state'} ) {
		$command .= "'  -S '".$reply->{'state'}."'";
	}
	if( defined $reply->{'locality'} ) {
		$command .= "'  -S '".$reply->{'locality'}."'";
	}
        system("$command -D '".$reply->{'domain'}."' -N 'schooladmin'");
        system("$command -D '".$reply->{'domain'}."' -N 'admin'");
        system("$command -D '".$reply->{'domain'}."' -N 'schoolserver'");
        system("$command -D '$CEPHALIX_DOMAIN' -N '".$reply->{'uuid'}."' -s");
}

#Create autoyast config

my $SCHOOL_DOMAIN   = $reply->{'domain'};
my $SCHOOL_sn       = $reply->{'uuid'};
my $block           = new Net::Netmask($reply->{'network'});
my $BITS            = $block->bits();

# In 4.0 it was changed:
$reply->{'network'}       = $block->base();
$reply->{'netmask'}       = $block->bits();
$reply->{'netmaskString'} = $block->mask();
my $dhcpBlock           = new Net::Netmask($reply->{'anonDhcpNetwork'});
my $dhcpLast = $dhcpBlock->broadcast();
my $serverBlock           = new Net::Netmask($reply->{'serverNetwork'});
$reply->{'serverNetworkmask'} = $serverBlock->bits();
if( $dhcpBlock->broadcast() eq $block->broadcast() ) {
   my @tmp  = split /\./, $dhcpBlock->broadcast();
   $dhcpLast = $tmp[0].'.'.$tmp[1].'.'.$tmp[2].'.'.( $tmp[3] - 1 );
}

$reply->{'anonDhcpRange'} = $dhcpBlock->base()." ".$dhcpLast;

# Evaluate WORKGROUP
my @ltmp = split /\./, $reply->{'domain'};
$reply->{'WORKGROUP'} = uc($ltmp[0]);

my $XML   = `cat $XMLFile`;
foreach my $par ( keys %{$reply} )
{
        my $val = $reply->{$par};
	Encode::_utf8_on($val);
	my $ph  = '###'.$par.'###';
        $XML =~ s/$ph/$val/g;
}

$SSLVARS{'REPLACE-SSHKEY'}     = `cat /root/.ssh/id_dsa.pub`;
$SSLVARS{'REPLACE-CA-CERT'}    = `cat $CEPHALIX_PATH/CA_MGM/cacert.pem`;
$SSLVARS{'REPLACE-VPN-KEY'}    = `cat $CEPHALIX_PATH/CA_MGM/certs/$SCHOOL_sn.$CEPHALIX_DOMAIN.key.pem`;
$SSLVARS{'REPLACE-VPN-CERT'}   = `cat $CEPHALIX_PATH/CA_MGM/certs/$SCHOOL_sn.$CEPHALIX_DOMAIN.cert.pem`;
$SSLVARS{'REPLACE-ADMIN-KEY'}  = `cat $CEPHALIX_PATH/CA_MGM/certs/admin.$SCHOOL_DOMAIN.key.pem`;
$SSLVARS{'REPLACE-ADMIN-CERT'} = `cat $CEPHALIX_PATH/CA_MGM/certs/admin.$SCHOOL_DOMAIN.cert.pem`;
$SSLVARS{'REPLACE-SCHOOL-KEY'} = `cat $CEPHALIX_PATH/CA_MGM/certs/schoolserver.$SCHOOL_DOMAIN.key.pem`;
$SSLVARS{'REPLACE-SCHOOL-CERT'}= `cat $CEPHALIX_PATH/CA_MGM/certs/schoolserver.$SCHOOL_DOMAIN.cert.pem`;

foreach my $sslpar ( @SSL )
{
        $XML =~ s/$sslpar/$SSLVARS{$sslpar}/g;
}
system("mkdir -p /srv/www/admin/{configs,isos}");
write_file("/srv/www/admin/configs/$SCHOOL_sn.xml",$XML);
write_file($Defaults,hash_to_json($default));

my $apache = "        ProxyPass          /$SCHOOL_sn http://".$reply->{'ipVPN'}."/api
        ProxyPassReverse   /$SCHOOL_sn http://".$reply->{'ipVPN'}."/api
";
write_file("/etc/apache2/vhosts.d/admin-ssl/$SCHOOL_sn.conf",$apache);
system("systemctl restart apache2");

my $tmpDir   = `mktemp /tmp/mkisoXXXXX`; chomp $tmpDir;
my $createCD = "cp /srv/www/admin/configs/$SCHOOL_sn.xml /usr/share/cephalix/templates/iso/oss.xml;\n".
	       "mkisofs --publisher 'Dipl.Ing. Peter Varkoly' -p 'CD-Team, http://www.cephalix.eu' ".
	       "-r -J -f -pad -no-emul-boot -boot-load-size 4 -boot-info-table -b boot/x86_64/loader/isolinux.bin ".
               "-c boot/boot.catalog -hide boot/boot.catalog -hide-joliet boot/boot.catalog -A - -V SU.001 -o /srv/www/admin/isos/$SCHOOL_sn.iso $tmpDir /usr/share/cephalix/templates/iso/ >/dev/null 2>&1";

system($createCD);

File: tools/create-institute-config.py
#!/usr/bin/python3
import json
import socket
import os
import sys
from ipaddress import *
#Define some defaults
# The file with the default values.
DEFAULTS_FILE = "/usr/share/cephalix/templates/Defaults.ini"
# Attributes which must not be changed
READONLY = ( "CEPHALIX", "CEPHALIX_PATH", "CEPHALIX_DOMAIN", "CCODE", "LANGUAGE", "NTP", "ZADMIN" )

#Read datas from stdin
institute=json.loads(sys.stdin.read())
for key in institute:
    try:
      institute[key] = institute[key].strip()
    except:
       pass
#Read the defaults from the config file
defaults=json.loads(open(DEFAULTS_FILE,"r").read())
SSLVARS = {}

CEPHALIX_PATH = defaults['CEPHALIX_PATH']

# The configuration files template file
xml_file          =  "/usr/share/cephalix/templates/autoyast-"+ institute.get("ayTemplate","default") + ".xml"
vpn_connect_file  =  "/usr/share/cephalix/templates/vpn-connect-institute-to-cephalix.sh"
full_connect_file =  "/usr/share/cephalix/templates/full-connect-institute-to-cephalix.sh"

#Set the read only values
for key in READONLY:
    institute[key] = defaults[key]

# create some networks
network        = IPv4Network(institute['internalNetwork'])
network_dhcp   = IPv4Network(institute['anonDhcpNetwork'])
network_server = IPv4Network(institute['serverNetwork'])
# calculate some network paramater
anon_dhcp_first = network_dhcp.network_address
anon_dhcp_last  = network_dhcp.network_address + network_dhcp.num_addresses -1
if network_dhcp.network_address == network.network_address: 
    anon_dhcp_first = anon_dhcp_first + 1
if anon_dhcp_last == network.network_address + network.num_addresses - 1:
    anon_dhcp_last = anon_dhcp_last - 1
institute['anonDhcpRange'] = "{} {}".format(anon_dhcp_first,anon_dhcp_last)
institute['netmask']       = network.prefixlen
institute['netmaskString'] = network.netmask
institute['internalNetwork'] = network.network_address
if 'WORKGROUP' not in institute:
    institute['WORKGROUP'] = institute['domain'].split('.')[0].upper()[0:15]

ident=institute['uuid']
if defaults.get('createIsoBy','uuid') == 'regCode':
    ident=institute['regCode']

# Handle VPN
if institute["ipAdmin"] != institute['ipVPN']:
    vpn_net = IPv4Network("{}/{}".format(institute['ipVPN'],'30'),False)
    with open("/etc/openvpn/ccd/"+ ident,"w") as ccd:
        ip_addresses = list(vpn_net.hosts())
        ccd.write("ifconfig-push {} {}\n".format(ip_addresses[0],ip_addresses[1]))
        if institute.get('fullrouting',False):
            ccd.write("iroute {} {}".format(network.network_address, network.netmask))
# Handle Certificates
cmd = CEPHALIX_PATH + '/create_server_certificates.sh -P ' + CEPHALIX_PATH + ' -O "' + institute['name'] + '"'
if 'state' in institute:
    cmd += ' -S "' + institute['state'] + '"'
if 'locality' in  institute:
    cmd += ' -L "' + institute['locality'] + '"'
if not os.path.isfile(CEPHALIX_PATH+'/CA_MGM/certs/' + ident +'.' + institute['CEPHALIX_DOMAIN'] + '.key.pem' ):
    command = cmd + ' -D ' + institute['CEPHALIX_DOMAIN'] + ' -N "' + ident + '" -s'
    os.system(command)
if not os.path.isfile(CEPHALIX_PATH+'/CA_MGM/certs/admin.' + institute['domain'] + '.key.pem' ):
    command = cmd + ' -D ' + institute['domain'] + ' -N admin'
    os.system(command)
if not os.path.isfile(CEPHALIX_PATH+'/CA_MGM/certs/cranix.' + institute['domain'] + '.key.pem' ):
    command = cmd + ' -D ' + institute['domain'] + ' -N cranix'
    os.system(command)
if not os.path.isfile(CEPHALIX_PATH+'/CA_MGM/certs/proxy.' + institute['domain'] + '.key.pem' ):
    command = cmd + ' -D ' + institute['domain'] + ' -N proxy'
    os.system(command)

SSLVARS['REPLACE-SSHKEY']     = open('/root/.ssh/id_rsa.pub','r').read()
SSLVARS['REPLACE-CA-CERT']    = open(CEPHALIX_PATH + 'CA_MGM/cacert.pem','r').read()
SSLVARS['REPLACE-VPN-KEY']    = open(CEPHALIX_PATH + 'CA_MGM/certs/' + ident + '.' + institute['CEPHALIX_DOMAIN'] + '.key.pem','r').read()
SSLVARS['REPLACE-VPN-CERT']   = open(CEPHALIX_PATH + 'CA_MGM/certs/' + ident + '.' + institute['CEPHALIX_DOMAIN'] + '.cert.pem','r').read()
SSLVARS['REPLACE-ADMIN-KEY']  = open(CEPHALIX_PATH + 'CA_MGM/certs/admin.' + institute['domain'] + '.key.pem','r').read()
SSLVARS['REPLACE-ADMIN-CERT'] = open(CEPHALIX_PATH + 'CA_MGM/certs/admin.' + institute['domain'] + '.cert.pem','r').read()
SSLVARS['REPLACE-SCHOOL-KEY'] = open(CEPHALIX_PATH + 'CA_MGM/certs/cranix.' + institute['domain'] + '.key.pem','r').read()
SSLVARS['REPLACE-SCHOOL-CERT']= open(CEPHALIX_PATH + 'CA_MGM/certs/cranix.' + institute['domain'] + '.cert.pem','r').read()
SSLVARS['REPLACE-PROXY-KEY']  = open(CEPHALIX_PATH + 'CA_MGM/certs/proxy.' + institute['domain'] + '.key.pem','r').read()
SSLVARS['REPLACE-PROXY-CERT'] = open(CEPHALIX_PATH + 'CA_MGM/certs/proxy.' + institute['domain'] + '.cert.pem','r').read()

institute['network']       = network.network_address
xml_content = open(xml_file,'r').read()
vpn_connect_content  = open(vpn_connect_file,'r').read()
full_connect_content = open(full_connect_file,'r').read()
for key in institute:
    value = "{}".format(institute[key])
    rkey  = '###'+key+'###'
    xml_content = xml_content.replace(rkey,value)
    vpn_connect_content  = vpn_connect_content.replace(rkey,value)
    full_connect_content = full_connect_content.replace(rkey,value)
for key in SSLVARS:
    xml_content = xml_content.replace(key,SSLVARS[key])
    vpn_connect_content = vpn_connect_content.replace(key,SSLVARS[key])
    full_connect_content = full_connect_content.replace(key,SSLVARS[key])
os.system('mkdir -p /srv/www/admin/{configs,isos}')


#write the config files
with open('/srv/www/admin/configs/' + ident + '.xml' ,'w') as f:
    f.write(xml_content)
with open('/srv/www/admin/configs/' + ident + 'vpn-connect.sh' ,'w') as f:
    f.write(vpn_connect_content)
with open('/srv/www/admin/configs/' + ident + 'full-connect.sh' ,'w') as f:
    f.write(full_connect_content)

#write apache configuration
with open('/etc/apache2/vhosts.d/admin-ssl/'+institute['uuid']+'.conf','w') as f:
    f.write("        ProxyPass          /{} http://{}/api\n".format(institute['uuid'],institute['ipVPN']))
    f.write("        ProxyPassReverse   /{} http://{}/api\n".format(institute['uuid'],institute['ipVPN']))
os.system('systemctl reload apache2')
os.system('/usr/share/cephalix/tools/create_institue_iso.sh {}'.format(ident))
